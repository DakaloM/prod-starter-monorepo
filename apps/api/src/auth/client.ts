import { BaseModel } from '@imax/datakit';

export enum ClientScope {
  auth = 'auth',
}

export class Client extends BaseModel {
  static tableName = 'client';
  id: string;
  secret: string;
  name: string;
  scope: string[];
}
