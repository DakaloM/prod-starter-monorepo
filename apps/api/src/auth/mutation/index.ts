export { login } from './login';
export * from './confirmEmail';