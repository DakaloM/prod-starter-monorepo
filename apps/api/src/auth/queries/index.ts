export { isClientAuthorized } from './isClientAuthorized';
export { isIdentityAuthorized } from './isIdentityAuthorized';
