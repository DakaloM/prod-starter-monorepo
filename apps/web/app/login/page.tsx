'use client';

import Image from 'next/image';
import Link from 'next/link';
import login_bg from '~/assets/login-bg.png';
import logo from '~/assets/logo.png';


export default function Login() {
  return (
    <section className="relative w-full h-screen bg-gray-100 flex items-center justify-center ">
      <Image 
      src={login_bg}
      alt='bg'
      className='absolute w-full h-full object-cover'
      />
      <div className='absolute w-full h-full bg-gray-100/95'>

      </div>
      <div className="w-full flex flex-col md:w-1/3 bg-white h-fit shadow-lg rounded-lg p-8 relative min-h-[200px]">
        <div className="flex items-center justify-between mb-8">
          <Image src={logo} className="w-16 h-auto object-cover" alt="logo" />

          <div className="flex flex-col gap-0 items-center">
            <h1 className="font-bold text-2xl">Welcome back!</h1>
            <p className="text-primary font-semibold text-xs">E-Recruitment</p>
          </div>
        </div>

        <div className="w-full px-10 grid gap-2">
          <div className="flex flex-col w-full items-start">
           
            <h6 className="font-normal text-sm text-gray-400 w-full text-center">Please login to your account</h6>
          </div>
  
          <div className="flex flex-col gap-2 text-center">
            <span className='text-sm font-normal text-gray-700'>
              Do not have an account?   
              <Link href='/register' className='text-primary ml-1 hover:underline'>Register here</Link>
            </span>
          </div>
        </div>
      </div>
    </section>
  );
}
