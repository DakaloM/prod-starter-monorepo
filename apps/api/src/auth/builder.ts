import { createBuilder } from '@imax/datakit';
import { faker } from '@imax/testkit';

import * as argon2 from 'argon2';
import 'argon2';
import { Client, ClientScope } from '~/auth/client';

export const clientBuilder = createBuilder(async (attrs: Partial<Client>, _factory, _ctx) => {
  const secret = attrs.secret || faker.internet.password(32);
  const encryptedSecret = await argon2.hash(secret);

  return Client.fromJson({
    ...attrs,
    name: faker.company.name(),
    secret: encryptedSecret,
    scope: [ClientScope.auth],
  });
});
