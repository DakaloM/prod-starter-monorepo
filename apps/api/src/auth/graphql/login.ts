import { builder } from '~/graphql/builder';

import { login } from '~/auth/mutation';

import { isAuthClient } from '../shield';
import { TokensRef } from './tokens';

builder.mutationField('login', (t) =>
  t.field({
    shield: isAuthClient,
    description: 'Authenticates a user',
    args: {
      email: t.arg.string({ required: true }),
      password: t.arg.string({ required: true }),
    },
    type: TokensRef,
    resolve: async (_root, args, ctx) => {
      return login(args, ctx);
    },
  }),
);

// builder.mutationField('refresh', (t) =>
//   t.field({
//     shield: isAuthClient,
//     description: 'Authenticates a user',
//     args: {
//       refreshToken: t.arg.string({ required: true }),
//     },
//     type: TokensRef,
//     resolve: async (_root, args, ctx) => {
//       return login(args, ctx);
//     },
//   }),
// );
