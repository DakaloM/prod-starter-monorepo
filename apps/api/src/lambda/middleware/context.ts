import { Handler, APIGatewayEventV2 } from '@imax/serverlesskit';

import middy, { MiddlewareObj } from '@middy/core';
import knex from 'knex';
import { AuthContext } from '~/auth';
import { DatabaseCache } from '~/cache/cache';
import { Context } from '~/context';
import { knexfile } from '~/knex';
import { mailer } from '~/mailer';
import { storageClient } from '~/storage';

const db = knex(knexfile);
export function addContext<TEvent extends ContextEvent>(): MiddlewareObj<TEvent> {
  return {
    before: async (request) => {
      const header = request.event.headers['authorization'] || '';
      const token = header.replace('Bearer ', '');

      const ctx = Context.init({
        db,
        mailer: mailer,
        cache: new DatabaseCache(db),
        storageClient,
      });

      const auth = await AuthContext.init(token, ctx);

      ctx.auth = auth;

      request.event.ctx = ctx;
    },
  };
}

export function withContext<TBody, TRes>(handler: Handler<TBody, TRes>) {
  return middy(handler).use(addContext());
}

export type ContextEvent = APIGatewayEventV2 & {
  ctx: Context;
};
