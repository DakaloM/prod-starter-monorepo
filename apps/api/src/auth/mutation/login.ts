import { ForbiddenError } from '@imax/serverkit';
import jwt from 'jsonwebtoken';

import * as argon from 'argon2';
import { User, UserStatus } from '~/account';
import { Context } from '~/context';

import { createAuthTokens } from './createAuthTokens';
import { config } from '~/config';

type CreateUserInput = {
  email: string;
  password: string;
};

export async function login(input: CreateUserInput, ctx: Context) {
  const user = await User.query(ctx.db).findOne({ email: input.email });

  if (!user?.password) {
    throw new ForbiddenError({
      message: 'Invalid credentials',
    });
  }

  if (user.status === UserStatus.Inactive) {
    throw new ForbiddenError({
      message: 'User account is inactive',
    });
  }

  if (user.status === UserStatus.NotConfirmed) {

    const secret = config.account.verifySecret || '154f60d2afd922d50bc4e384069c5f8267dbf0dec4ced6eeeb52eadd5dd07ffb';
    const token = jwt.sign({ id: user.id, email: input.email }, secret, { expiresIn: '1d' });

    await ctx.mailer.sendMail({
      to: input.email,
      subject: 'Confirm your email',
      data: {
        title: 'Confirm your email',
        subTitle: 'Confirm your email address',
        paragraphs: [
          `Please click the link below to confirm your email`,
          `<a href="http://localhost:3000/confirm-email?token=${token}">Confirm email</a>`,
        ],
        recipientName: `${user.name} ${user.surname}`,
      },
    });

    throw new ForbiddenError({
      message: 'Your email address has not been confirmed, please check your email for instructions',
    });
  }

  const isValid = await argon.verify(user.password, input.password);

  if (!isValid) {
    throw new ForbiddenError({
      message: 'Invalid credentials',
    });
  }

  return createAuthTokens(user, ctx);
}
