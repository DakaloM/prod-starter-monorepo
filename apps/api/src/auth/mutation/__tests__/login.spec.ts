import { faker } from '@imax/testkit';

import crypto from 'node:crypto';
import { Token } from '~/auth/token';
import { Context } from '~/context';
import { createContext, destroyContext } from '~/test/context';
import { Environment, createEnv, destroyEnv } from '~/test/environment';
import { Factory, createFactory } from '~/test/factory';

import { login } from '../login';

describe('auth/mutation/login', () => {
  let env: Environment;
  let ctx: Context;
  let factory: Factory;
  beforeAll(async () => {
    env = await createEnv();
  });

  afterAll(async () => {
    await destroyEnv(env);
  });

  beforeEach(async () => {
    ctx = await createContext(env);
    factory = createFactory(ctx);

  });

  afterEach(async () => {
    await destroyContext(ctx);
  });

  it('should throw error when user does not exist', async () => {
    await expect(
      login(
        {
          email: faker.internet.email(),
          password: faker.internet.password(),
        },
        ctx,
      ),
    ).rejects.toMatchSnapshot();
  });

  it('should throw error when user does not have a password', async () => {
    const user = await factory.insert('user');
    await expect(
      login(
        {
          email: user.email,
          password: faker.internet.password(),
        },
        ctx,
      ),
    ).rejects.toMatchSnapshot();
  });

  it('should authenticate user', async () => {
    const password = crypto.randomBytes(10).toString('hex');
    const user = await factory.insert('user', { password });

    const tokens = await login({ email: user.email, password }, ctx);

    expect(tokens.accessToken).toBeDefined();
    expect(tokens.refreshToken).toBeDefined();
    expect(tokens.expiresIn).toBeDefined();
    expect(tokens.expiresAt).toBeDefined();

    const token = (await Token.fromString(tokens.accessToken, ctx)) as Token;
    expect(token).toBeDefined();
    expect(token.userId).toEqual(user.id);
  });
});
