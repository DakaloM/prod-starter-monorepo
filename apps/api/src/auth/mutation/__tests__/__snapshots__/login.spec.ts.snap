// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`auth/mutation/login > should throw error when user does not exist 1`] = `[Error: Invalid credentials]`;

exports[`auth/mutation/login > should throw error when user does not have a password 1`] = `[Error: Invalid credentials]`;
